File: mirea/cources/beginner/connected-components-3/b.py
# структура для системы непересекающихся множеств
class DSU:
    
    def __init__(self, n):
        self.sz = [1] * n;
        self.par = [int(x) for x in range(n)];
    
    def getPar(self, u):
        if u == self.par[u]:
            return u;
        p = self.getPar(self.par[u]);
        self.par[u] = p;
        return p;
    
    def connect(self, u, v):
        u, v = self.getPar(u), self.getPar(v);
        if u != v:
            if self.sz[u] < self.sz[v]:
                u, v = v, u; # swap
            self.sz[u] += self.sz[v];
            self.par[v] = u;

# оставить только уникальные элементы в сортированном списке
def unique(s):
    i, p, n = 0, 0, len(s);
    while (i < n):
        j = i;
        while (j < n and s[i] == s[j]):
            j += 1;
        s[p] = s[i];
        i = j;
        p += 1;
    while (n > p):
        s.pop();
        n -= 1;

# создаём систему непересекающихся множеств для 26 букв
dsu = DSU(26);

# читаем данные. каждый ключ соединяет рёбрами буквы, которые
# в него входят. проводим эти рёбра
n = int(input());
used = [False] * 26;
for i in range(n):
    # читаем ключ, сортируем буквы и удаляем повторы
    s = sorted(input());
    unique(s);
    # отмечаем, какие буквы были использованы
    k = len(s);
    for i in range(k):
        used[ord(s[i])-ord('a')] = 1;
    # проводим рёбра
    for i in range(k-1):
        dsu.connect(ord(s[i])-ord('a'), ord(s[i+1])-ord('a'));

# выводим количество компонент связности среди задействованных
# во входных данных латинских букв
parents=[]
for i in range(26):
    if used[i]:
        parents.append(dsu.getPar(i));
parents.sort();
unique(parents);
print(len(parents));

File: mirea/cources/beginner/connected-components-3/b.cpp
#include <bits/stdc++.h>
#define all(x) (x).begin(),(x).end()
#define isz(x) (int)(x).size()
using namespace std;
// структура данных система непересекающихся множеств с ранговой
// эвристикой и эвристикой сжатия путей
struct DSU {
    vector<int> par, sz;
    DSU(int n) : par(n), sz(n, 1)
    {
        for (int i = 0; i < n; i++)
            par[i] = i;
    }
    int getPar(int u) {
        return par[u] == u ? u : par[u] = getPar(par[u]);
    }
    void connect(int u, int v) {
        u = getPar(u), v = getPar(v);
        if (u == v)
            return;
        if (sz[u] < sz[v])
            swap(u, v);
        sz[u] += sz[v];
        par[v] = u;
    }
};
void solve() {
    // всего может быть не больше 26 компонент - по одной
    // на каждую букву
    // один ключ объединяет в компоненту все буквы, которые содержит
    int n; cin >> n;
    DSU dsu(26);
    vector<bool> used(26);
    for (int i = 1; i <= n; i++) {
        // читаем строку
        string s; cin >> s;
        // сортируем элементы и удаляем повторы
        sort(all(s));
        s.erase(unique(all(s)), s.end());
        // отмечаем какие буквы вообще были использованы в данном тесте
        for (char c : s)
            used[c-'a'] = 1;
        // соединяем соседние буквы ключа
        for (int j = 1; j < isz(s); j++)
            dsu.connect(s[j-1]-'a', s[j]-'a');
    }
    // выводим ответ, получая количество различных компонент связности
    set<int> parents;
    for (int i = 0; i < 26; i++)
        if (used[i])
            parents.insert(dsu.getPar(i));
    cout << isz(parents) << endl;
}
main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int tt = 1; //cin >> tt;
    while (tt --> 0)
        solve();
}
